### PR TITLE
Adds shape testing with streams

### PR DESCRIPTION
**Pull request by samjonester**

### CODE REVIEW

File: app/posts/2017-07-21-shape-testing-with-javascript-streams.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+
+module.exports = function validReport(report) {
+  return allPass([
+      hasFields(reportFields),
+      anyAt(violationsPath, hasFields(violationFields))
+    ], report);
+}
+```
+
```
Comment (stevenjackson) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: This code feels "dense" to me.  Have you experimented with explaining it in smaller chunks?  I think this is neat and powerful, but I got distracted trying to figure out what the validation rules were in the middle.

File: app/posts/2017-07-21-shape-testing-with-javascript-streams.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// => new partially applied function waiting for data
+const plusOne = map(n => n+1)
+const waitedForData = plusOne([1, 2, 3])
+
+// allAtOnce === waitedForData
+```
+
+### Composing functions
+
+A composed function is just multiple functions glued together. The trick is to compose functions that have been partially applied, or are still waiting on their thing to operate on. The entire composition will then be a pipeline, waiting on the thing that the first function is also waiting on.
```
Comment (stevenjackson) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: I like these explanations for composition and currying :+1:  Really helps connect the dots on how pipelines work.

File: app/posts/2017-07-21-shape-testing-with-javascript-streams.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+module.exports = function validReport(report) {
+  return allPass([
+      hasFields(reportFields),
+      anyAt(violationsPath, hasFields(violationFields))
+    ], report);
+}
+```
+
+Something really cool has happened by writing our validations this way. To understand the purpose of `validReport` we can just read the function names, we don't need to translate "how the functions operate" into "what is the purpose of this code". That's the most powerful concept to understand when getting started with functional programming.
+
```
Comment (stevenjackson) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: I feel like it ended too soon.  I almost expect the conclusion to reference shape testing again just to remind me how we ended up down here.

Comment (samjonester) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: Good point. I've added a conclusion to tie it all back together. Thanks!

File: app/posts/2017-07-21-shape-testing-with-javascript-streams.md (hunk 1 above)
Comment (samjonester) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: Thanks, I think you're right. I separated it out into two pieces. The first to define the shape of a valid report, and the second to create the validation utilities.

File: app/posts/2017-07-21-shape-testing-with-javascript-streams.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+const any = require('lodash/fp/any')
+const compose = require('lodash/fp/compose')
+const get = require('lodash/fp/get')
+const has = require('lodash/fp/has')
+const map = require('lodash/fp/map')
+
+module.exports.hasfields = compose(allPass, map(has))
+module.exports.anyAt = (path, check) => compose(any(check), get(path))
+```
+What we've done is use function composition to glue together smaller things into utilities that check that an object has values for given fields with `hasFields`, and that an object passes a validation check for a certain leaf node.
```
Comment (samjonester) [inline app/posts/2017-07-21-shape-testing-with-javascript-streams.md]: Values is such an ambiguous term. I'm ok with it, I think.

**Pull request merged**